### ISSUE TITLE
Support multiple storage files in different locations

### ISSUE THREAD

**Issue by dumblob** [Tomas-M/dynfilefs#1]
### Motivation

1. use multiple partitions on multiple physical disks (USB, NVME, SATA, ...) to store large data (videos, etc.)
2. overcome VFAT limitation of ~4GByte per storage file
    a. to allow for saving say pictures (~7MByte each) amounting to >4 GByte size
    b. to allow for saving say 4K videos (~150GByte each) amounting to tens of terabytes of storage

### Solution

Support multiple storage files in different locations. The order on command line could set their priority for "bucket filling". Large files (not fitting in whole to any of the free spaces on either of the multiple storage files) could be split by binary division (this is to keep as much as possible of the original data contiguous - good for data recovery, for speed as apps often expect this, etc.).

Another cool trick would be to store files/chunks in parallel to each of the storage files using a "work stealing" algorithm (which automatically accounts for e.g. having two storage files on the same physical disk) but I think it would be an unnecessary complication code-wise with not so much speed gains in practical scenarios (but this needs testing).

**Comment by Tomas-M** [comment]
This is interesting idea.

The virtual.dat file size needs to be known at the beginning and it is impossible to change it later. So you would need to know, beforehand, how big storage do you want to provide. Also the index of offsets needs to be allocated beforehand as well, so for example if you plan to provide 1TB virtual.dat file, it will require to allocate 2GB of index data beforehand (on the first write).

This is not practical in my opinion.
However, having the possibility to split the storage to multiple files and overcome the 4GB single-file limitation is something I will consider, this seems beneficial to me.

**Comment by dumblob** [comment]
> The virtual.dat file size needs to be known at the beginning and it is impossible to change it later. So you would need to know, beforehand, how big storage do you want to provide.

That is totally fine. There are many strategies one can cope with this.  This multiple strategies stem from the fact than one has already a set of regular "files" which can be arbitrarily copied & moved (assuming the linking scheme *dynfile* uses can accommodate for path/name chages - which it should anyway due to volatility of USB flash disks naming etc.).

>Also the index of offsets needs to be allocated beforehand as well, so for example if you plan to provide 1TB virtual.dat file, it will require to allocate 2GB of index data beforehand (on the first write).
>
> This is not practical in my opinion.

I am happy to pay this price - it is a one-time slowdown and if one already has 1TB of space, then the writing of 2GB of data contiguously would most probably be fast enough (few seconds on **average** SSDs with ~500 MByte/s write speeds).

What do you think?

>However, having the possibility to split the storage to multiple files and overcome the 4GB single-file limitation is something I will consider, this seems beneficial to me.

Thanks! :+1:

**Comment by Tomas-M** [comment]
I will think about it. But I do not promise anything.
If you have so huge data amounts, I would suggest to use different filesystem than FAT32.
Linux provides many different solutions to join storage spaces from different sources, DynFileFS is not ideal for this.

**Comment by dumblob** [comment]
> I will think about it. But I do not promise anything.

Sure, do not feel pressured. If you need any help, please reach out - C and UNIX is in my veins since almost the beginning of Slax :wink:.

>If you have so huge data amounts, I would suggest to use different filesystem than FAT32. Linux provides many different solutions to join storage spaces from different sources, DynFileFS is not ideal for this.

That is true. I use(d) a few and they have different trade-offs which turn out to be obstackles for me (and not just me judging based on issue tracker tickets of those projects). So DynFileFS seems to be another alternative but with its simplicity actually potentially surpassing other more complicated SW.

There are also other advantages important to me - e.g. it is much easier to set up bit rot protection on files (i.e. in user space) than on whole file systems. This also helps a lot if there in case of bad USB connection or SATA cable etc. So DynFileFS could occupy a sweet spot in this space. Will see how it will turn out :wink:.

-----
Btw. speaking of FAT32, any plans to switch to "true" VFAT with Slax (if you did not already)? Linux has now native support which is getting continually improved.

**Comment by Tomas-M** [comment]
I am not sure what you mean by true vfat, Slax uses "vfat" kernel driver to mount fat volumes already.

**Comment by dumblob** [comment]
Sorry, tired today and wrote vfat instead of exfat. I meant [exFAT](https://en.wikipedia.org/wiki/ExFAT ).

**Comment by Tomas-M** [comment]
The problem with exFAT is that I am unable to boot from it, I do not know any tools which would boot from it. Furthermore if the computer wants to boot with UEFI, it seems it requires separate FAT32 partition anyway. Furthermore I am not aware that exFAT driver would be in linux kernel, seems like there is only userspace support with FUSE

**Comment by dumblob** [comment]
> The problem with exFAT is that I am unable to boot from it, I do not know any tools which would boot from it. Furthermore if the computer wants to boot with UEFI, it seems it requires separate FAT32 partition anyway.

Actually UEFI does not support FAT32 per spec. But in practice many seem to do - same as they seem to switch to BIOS booting if not successful in booting using UEFI mechanisms (specced or not) and thus "support" exFAT - see https://superuser.com/a/1025445 . All in all it could "just work" in most cases. Might be worth trying as some low-priority fun activity - see below.

>Furthermore I am not aware that exFAT driver would be in linux kernel, seems like there is only userspace support with FUSE

That is the "new thing" I had in mind - the current mainline vanilla kernels do have **full** support for exFAT. This mainline code is even being used by Samsung in their HW (incl. phones), so it can be trusted (to a degree as with any fs :wink:). No FUSE involved :wink:.

**Comment by Tomas-M** [comment]
So my main objective now is to support splitting of storage file into smaller pieces to overcome the FAT filesystem limitation per single file (which is 4GB minus 1 byte).

**Comment by Tomas-M** [comment]
You can test the current master branch. I've implemented -p parameter, which allows you to specify split size for your storage file. So you can have 16GB virtual file for example and split it into 4000MB pieces. Feel free to test and report any problems you may encounter.
